File: selection/_041/Main.java
package selection._041;

import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int d = sc.nextInt();
        int n = sc.nextInt();
        int[] t = new int[d];
        int[] a = new int[n];
        int[] b = new int[n];
        int[] c = new int[n];
        for (int i = 0; i < d; i++) {
            t[i] = sc.nextInt();
        }
        for (int i = 0; i < n; i++) {
            a[i] = sc.nextInt();
            b[i] = sc.nextInt();
            c[i] = sc.nextInt();
        }
        sc.close();

        // dp[i][j]: i日目にj番目の服を選んだときのi日目までの最大値
        int[][] dp = new int[d + 1][n + 1];
        for (int i = 2; i <= d; i++) {
            // i日目に着る服
            for (int j = 1; j <= n; j++) {
                // i - 1日目に着る服
                for (int j2 = 1; j2 <= n; j2++) {
                    // i日目に着る服が気温の条件を満たすか
                    if (!(a[j - 1] <= t[i - 1] && t[i - 1] <= b[j - 1])) {
                        continue;
                    }
                    // i - 1日目に着る服が気温の条件を満たすか
                    if (!(a[j2 - 1] <= t[i - 2] && t[i - 2] <= b[j2 - 1])) {
                        continue;
                    }
                    dp[i][j] = Math.max(dp[i - 1][j2] + Math.abs(c[j - 1] - c[j2 - 1]), dp[i][j]);
                }
            }
        }

        int result = 0;
        for (int i = 1; i <= n; i++) {
            result = Math.max(result, dp[d][i]);
        }

        System.out.println(result);
    }
}

File: selection/_041/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	var d, n int
	fmt.Scan(&d, &n)

	t, a, b, c := make([]int, d), make([]int, n), make([]int, n), make([]int, n)
	for i := 0; i < d; i++ {
		fmt.Scan(&t[i])
	}
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i], &b[i], &c[i])
	}

	dp := make([][]int, d+1)
	for i := 0; i < d+1; i++ {
		dp[i] = make([]int, n)
	}

	// dp[i][j]: i日間まででi-1日目に服jを着るときの派手さの合計
	// dp[i][j] → dp[i+1][k] (k:0~n-1)
	for i := 1; i < d; i++ {
		for j := 0; j < n; j++ {
			for k := 0; k < n; k++ {
				if !(a[j] <= t[i-1] && t[i-1] <= b[j]) {
					continue
				}
				if !(a[k] <= t[i] && t[i] <= b[k]) {
					continue
				}

				dp[i+1][k] = int(math.Max(float64(dp[i+1][k]), float64(dp[i][j]+int(math.Abs(float64(c[j]-c[k]))))))
			}
		}
	}

	result := 0
	for i := 0; i < len(dp[d]); i++ {
		result = int(math.Max(float64(result), float64(dp[d][i])))
	}

	fmt.Println(result)
}
